### ISSUE TITLE
Puppeteer not working when is not focused

### ISSUE THREAD

**Issue by JaLe29** [puppeteer/puppeteer#3339]
**Environment:**

* Puppeteer version: 1.8.0
* Platform / OS version: Win 10 64b and Ubuntu 18.04.1 LTS
* URLs (if applicable): No
* Node.js version: v10.6.0
 
I am using puppeteer for testing my pages. When I run the app and hide it into background then puppeteer is paused (or freeze). When I move puppeteer into foreground is running normally.

I think that its problem "when window is not focused, its not working".

This behavior is on windows and linux too.

Set up code:

```
			const args = []
			args.push('--start-maximized')
			args.push('--disable-gpu')
			args.push('--disable-setuid-sandbox')
			args.push('--force-device-scale-factor')
			args.push('--ignore-certificate-errors')
			args.push('--no-sandbox')

			this.driver = await puppeteer.launch({
				headless: false,
				args,
				userDataDir: `puppeteer_profile`,
			})


```

i happen only "sometimes" I don't know when.

**Comment by aslushnikov** [comment]
@JaLe29 what do you mean by "not running"? Does the page's javascript stop running?

**Comment by JaLe29** [comment]
Yes, page`s javascript are **paused**. While window is not focused.

I think, this behaviour is related something settings in my operating system.

But as I am writing in first post, i tested on two OS - linux and windows.

**Comment by aslushnikov** [comment]
@JaLe29 This is probably related to  https://github.com/WICG/web-lifecycle/

The easy workaround would be:
```js
const session = await page.target().createCDPSession();
await session.send('Page.enable');
await session.send('Page.setWebLifecycleState', {state: 'active'});
```

This should unfreeze javascript execution in the background page.

**Comment by entrptaher** [comment]
@aslushnikov when we have multiple pages, will adding this snippet make all of them focused? Also it's `page.target()` as documented.
I tried with the snippet, 
```js
const puppeteer = require("puppeteer");

(async () => {
  const browser = await puppeteer.launch({ headless: true });
  const page = await browser.newPage();
  
  const session = await page.target().createCDPSession();
  await session.send("Page.enable");
  await session.send("Page.setWebLifecycleState", { state: "active" });

  await page.goto("http://127.0.0.1:8080/");
  await page.screenshot({ path: "example.png", fullPage: true});

  await browser.close();
})();
```

and the html code,
```html
<script type="text/javascript">
  window.onload = function () {
    function check() {
      if (document.hasFocus() == lastFocusStatus) return;
      lastFocusStatus = !lastFocusStatus;
      statusEl.innerText = lastFocusStatus ? 'with' : 'without';
    }
    window.statusEl = document.getElementById('status');
    window.lastFocusStatus = document.hasFocus();
    check();
    setInterval(check, 200);
  }
</script>

Currently <b id="status">without</b> focus...
```

The result,
![image](https://user-images.githubusercontent.com/8284972/46604117-f9795e80-cb16-11e8-954d-65cc05771452.png)

It didn't focus back. Should it focus back?

**Comment by JaLe29** [comment]
@aslushnikov I tested your solution

```
const session = await page.target().createCDPSession();
await session.send('Page.enable');
await session.send('Page.setWebLifecycleState', {state: 'active'});
```

But it not working too. Still same behaviour. :-(

**Comment by JaLe29** [comment]
I'm the only one who has this problem?

With this error it's useless.

**Comment by entrptaher** [comment]
Nope you are not only one, I have same problem above.

**Comment by JaLe29** [comment]
My friend has same problem. I don't have any idea how to fix it. Sometimes is working, sometimes not.

I must focus manually the window and then it's working.

**Comment by jonesprabu** [comment]
I have a similar issue, In VM i have scheduled to execute the puppeteer stuffs using node-corn in my nodejs app. I disconnected the VM and connected after some hrs of the actual scheduled time i can see the log as the scheduler has started at the correct time, i havnt seen any log entry in the puppeteers code, then i enter in the cmd window, then the puppeteer stuffs run, u can see the time after the job started log entry. 

![image](https://user-images.githubusercontent.com/3309381/48718494-84cf2f00-ec41-11e8-8e59-4d82c4545d9e.png)


`
```
import puppeteer = require('puppeteer');
import reader = require('./webPageReader');
import config = require('./config');

import { configure, getLogger } from 'log4js';
configure(config.project.LOG_CONFIG);
const logger = getLogger('bdayWisherLog');


import cron = require('node-cron');

let start = async () => {
    logger.debug("Job Fired...");
    const browser:puppeteer.Browser = await puppeteer.launch({headless: true});
    reader.runThroughPage(browser);
};

let startSchedule = () => {
    logger.debug('Schedule started... Job will fire by '+config.schedule.HR+':'+ config.schedule.MIN);

    process.on("unhandledRejection", (reason, p) => {
        logger.debug("Unhandled Rejection at: Promise", p, "reason:", reason);
    });

    cron.schedule(config.schedule.MIN+' '+ config.schedule.HR +' * * *', ()=>{
        start();
    }, {
        scheduled: true,
        timezone: "Asia/Kolkata"
    });
}

startSchedule();
```
`

**Comment by JaLe29** [comment]
@jonesprabu **Totally** the same problem.  I am using a scheduling system too. But I am using `setTimeout `for start new task.

**Comment by JaLe29** [comment]
@aslushnikov any updates about this? I am so sad that I can't use this amazing lib :(

**Comment by aslushnikov** [comment]
@JaLe29 can you please try running the following script and confirm that "heartbits" stop from both pages once you minimize browser window?

```js
const puppeteer = require('.');

(async() => {
  const browser = await puppeteer.launch({
    headless: false,
  });
  const [page] = await browser.pages();
  await setupPageHeartbit('ONE', page);
  await setupPageHeartbit('TWO', await browser.newPage());
})();

async function setupPageHeartbit(pageId, page) {
  page.on('console', msg => console.log(`[${pageId}]: ${msg.text()}`));
  await page.evaluate(() => {
    let i = 0;
    setInterval(() => {
      console.log((++i) + ' heartbit');
    }, 1000);
  });
}
```

**Comment by abhi4singhal** [comment]
> @aslushnikov I tested your solution
> 
> ```
> const session = await page.target().createCDPSession();
> await session.send('Page.enable');
> await session.send('Page.setWebLifecycleState', {state: 'active'});
> ```
> But it not working too. Still same behaviour. :-(

@JaLe29 I used the solution mentioned by @aslushnikov and it solved the problem.
I used this snippet for every tab that I was creating and the JS execution is no longer halted.

**Comment by MatteoGioioso** [comment]
Same issue here, I have 2 tabs and the tab that is not focused pauses. And the session tricks does not work for me

**Comment by cyphx** [comment]
I also faced the same problem. I tried the snippet given by @aslushnikov , but it didn't work. I read in a report where they advised running on multiple instances of browser instead of trying to manage different tabs and it worked for me. It does not pause in separate browser instances.
So, if you are stuck and your requirement allows you to use/test multiple browser instances, I think you should go for it.

**Comment by Alex0007** [comment]
If you can't use multiple browser instances, you can stick to some previous version of puppeteer

![image](https://user-images.githubusercontent.com/1143226/50427037-b7aeb780-08be-11e9-8597-01cbbbf815db.png)

I randomly tested 1.3.0 version and there is no tab discarding exist in chromium

**Comment by ghost** [comment]
I'm having somewhat a similar problem. I'm launching 5 Chrome 73.0.3683.103 instances (not chromium) on Centos 7. I launch them as child processes so they all run concurrently no problem.

If I don't minimize them, they run fine even if they are not focused. If I run a script to miminize the windows, they freeze execution until I unminimize. I'm not sure if this is website specific or a problem with chrome/puppeteer. I don't have time to test more extensively, just wanted to share my experience.

**Comment by fgarciajulia** [comment]
I have the same issue.

**Comment by VikramTiwari** [comment]
Hey @aslushnikov I have tested the script you provided above and heartbits do not stop. They do keep happening.

But when I try running normal script, it goes into the paused state and nothing happens on that website page. Things also don't get timed out from puppeteer's side. It's as if the whole puppeteer script and browser just went into a paused state.

I read the WICG/page-lifecycle which goes into more details about the frozen state. But then how come this only happens when I am trying to execute something via puppeteer but not in normal chrome. Or is the argument that it happens in normal execution as well but we don't see it because we are not actively trying to do anything in that browser.

Shouldn't this be treated as bug for chromium though. If script is trying to do something in the current page, then that page should never get into the frozen state. If this is already an open bug, could someone link it please. Thanks!

Also, this didn't help: https://github.com/GoogleChrome/puppeteer/issues/3339#issuecomment-426907894

**Comment by martijn189** [comment]
This is also a problem for me, is there any news or information on how to fix this? If i run Puppeteer in "headless:true" mode it seems to work just fine. So it only happens to me when the GUI is used (minimize browser pauses script, switching tabs pauses script

**Comment by bryanmacnz** [comment]
Also having this problem i.e. with multiple tabs open only the tab that has focus continues to work properly. Pages without focus "stop running" and '.waitForSelector' calls time out. Very keen for a fix.

**Comment by johnnyhappy365** [comment]
I have same problem

**Comment by lookitsatravis** [comment]
This is happening to me as well. Specifically, `page.type()` will pause until the non-headless Chrome window regains focus. I have a cron which runs every 10 minutes, and if I minimize or switch windows, the typed text will not complete until I restore the window.

**Comment by david-littlefield** [comment]
### [Solution]:

Add the following values to the array you assign to the args property in the options object. Now, you should be able to run Puppeteer while your window is in the foreground, background, or minimized. 

**Update:**
I'm running parallel instances of Puppeteer using the same script in multiple windows with one tab per window. Each window completed successfully whether it was in the foreground, background, or minimized. And, it worked in both headless and headful mode.

```
const chromeArgs = [
  '--disable-background-timer-throttling',
  '--disable-backgrounding-occluded-windows',
  '--disable-renderer-backgrounding'
];
```
Solution and explanation available in an article written by **Rohit Kumar**, @rkr090 at Big Binary:

[Debugging failing tests in puppeteer because of background tab](https://blog.bigbinary.com/2018/08/15/debugging-failing-tests-in-background-tab-in-puppeteer.html)

**Comment by VikramTiwari** [comment]
Hey @captaindavepdx The issue is still there as mentioned here: https://github.com/GoogleChrome/chrome-launcher/issues/169

**Comment by david-littlefield** [comment]
Hi @VikramTiwari ! Have you tried using multiple windows instead of multiple tabs?

**Comment by VikramTiwari** [comment]
I haven't. Multiple windows only work if the complete script execution is limited to a single page. As soon as execution is multiple page, we will face the issue again.

**Comment by david-littlefield** [comment]
Could you please help me understand why multiple windows would only work for single page execution?

**Comment by VikramTiwari** [comment]
I think for me the UX is very icky if done with utilizing multiple windows rather than tabs.

An example would be: 
- Open Page A
- Click on a link to open Page B in a new tab/window
- Click on a button on Page B
- Some content on Page A changes
- Copy the content from Page A

Do this 50 times and we end up with 100 chrome windows taking over desktop. But if done with tabs, you have everything as tabs. Much cleaner UX.

Also, a solution would let script execute when tab or window is not focused. This strategy is a patch workaround to achieve a solution with windows. Nothing against the patch, it just isn't the solution.

**Comment by david-littlefield** [comment]
Thanks for that clarification. I can see how it could be considered a patch or workaround for your situation. But, since the OP didn't mention multiple pages, maybe it could still serve as as a solution for others.

Also, the patch or workaround does allow the script to execute while the window is not focused.

**Comment by entrptaher** [comment]
I have to run 100+ puppeteer windows (not tabs) in parallel with `puppeteer.launch()`. If I have to use this workaround on all 100 windows, it will be a disaster for me. :(

**Comment by captaindavepdx** [comment]
Hi @entrptaher, 100+ puppeteer windows, that's awesome! Would you mind sharing your setup?

**Comment by Nisthar** [comment]
I am having the same issue. I am using `puppeteer-core` and `connect` instead of launch.

**Comment by hubitor** [comment]
Same issue here...
Sometimes when I run a scrape script, puppeteer opens the browser and puts focus on it. If I click on another window the script pauses. When I click back to the browser window of puppeteer the script continues execution.

OS: Arch Linux

**Comment by ganyanchuan1989** [comment]
i have the same problem。 set handless = false fix problem。

but handless = false debug is not friendly

**Comment by Nisthar** [comment]
@ganyanchuan1989 I am using `puppeteer.connect` with headless: false. Its still not working for me.

**Comment by hejiheji001** [comment]
Same issue, so this problem lasts for more than a year??

**Comment by wonghang** [comment]
I am suffering the same issue. Any fix or workaround?

**Comment by alansyue1020** [comment]
it works on headless : false mode, but if i use headless: true, this problem would happen

**Comment by wonghang** [comment]
Yes. I need headless: false

**Comment by SunshowerC** [comment]
same here.

**Comment by ecolss** [comment]
> I also faced the same problem. I tried the snippet given by @aslushnikov , but it didn't work. I read in a report where they advised running on multiple instances of browser instead of trying to manage different tabs and it worked for me. It does not pause in separate browser instances.
> So, if you are stuck and your requirement allows you to use/test multiple browser instances, I think you should go for it.

@abhi4singhal Yes, running multiple browsers (other than pages) really worked.
But this also brings a new problem, that is, the urls I need to scrape requires some init work.
For example if the browser launched open a url for the first time, the website shows some welcome messages that I need to manually interact with (such as `accept cookie`, `choose location` blabla).

With 1 browser + multiple pages, I could manually do this init work for once (to create user_data for those websites), later I won't have the init work for this particular browser (as long as I don't clean the user_data).

But with multiple browsers, this init work cannot be done manually, any advice?

**Comment by steve-allied-man** [comment]
Same here.
@aslushnikov do you guys have any update on this or this not something you expect to work? Hope it will be resolved once you have a chance. Cheers :)

**Comment by joaovmvini** [comment]
same issue... i really need a solution

**Comment by yourbuddyconner** [comment]
> ### [Solution]:
> Add the following values to the array you assign to the args property in the options object. Now, you should be able to run Puppeteer while your window is in the foreground, background, or minimized.
> 
> **Update:**
> I'm running parallel instances of Puppeteer using the same script in multiple windows with one tab per window. Each window completed successfully whether it was in the foreground, background, or minimized. And, it worked in both headless and headful mode.
> 
> ```
> const chromeArgs = [
>   '--disable-background-timer-throttling',
>   '--disable-backgrounding-occluded-windows',
>   '--disable-renderer-backgrounding'
> ];
> ```
> 
> Solution and explanation available in an article written by **Rohit Kumar**, @rkr090 at Big Binary:
> 
> [Debugging failing tests in puppeteer because of background tab](https://blog.bigbinary.com/2018/08/15/debugging-failing-tests-in-background-tab-in-puppeteer.html)

FWIW, this solution  worked for me today. I am scraping via an existing chrome session and connecting using `Puppeteer.connect`. I was having the same issue where the out-of-focus Chrome window would not execute Javascript in the background. 

This is the command I used to start the chrome instance: 

`/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --remote-debugging-port=9222 --no-first-run --no-default-browser-check --disable-backgrounding-occluded-windows --disable-renderer-backgrounding  --disable-background-timer-throttling --user-data-dir=$(mktemp -d -t 'chrome-remote_data_dir')`

Note the use of the flags: `--disable-backgrounding-occluded-windows --disable-renderer-backgrounding  --disable-background-timer-throttling`

For those having issues, this seems to be a feature of Chrome/Chromium/browsers in general that is designed to protect the user from websites snooping on them while they're not actively using the website. This feature can and should be turned off in the context of a web scraper running via a scheduler, headless or otherwise.

**Comment by HongLiangGong1990** [comment]
same issue... puppeteer script paused when switch to another tab, i need multi tabs to work simultaneously. but i have yet to find good solution now. help!!!!

**Comment by Solaflex** [comment]
Same issue with around 45 browser instances parallel open. The three Chrome command line arguments are not solving the issue.

**Comment by OrenSchwartz** [comment]
any news ?

**Comment by theArina** [comment]
any news ?

**Comment by bachega8** [comment]
I get this problem when i try to perform some click in the page.  I solved this using a diferent way 
to perform clicks. Check out:

Before:
```js
let login_button = await page.$x('//a[contains(text(), "Login")]')  
 await login_button[0].click()
```

After:

```js
await page.evaluate(() => {
        let but = document.querySelector("span[class='pr-2 pr-sm-2 pr-lg-4 pl-sm-2 pl-md-2 pt-1 pb-1 nao_logado active'] > a")
        if (but) {
            return Promise.resolve(but.click())
        }
    })
```

I don't know why, but using "evaluate()" i can click on page buttons and links with background, minimized mode

**Comment by paolomangiadev** [comment]
+1

**Comment by 1nVitr0** [comment]
Also still experiencing this issue. Any amount of flags passed to chrome have absolutely no effect (on Microsoft Windows). Will try to reproduce a minimal working example. We have 8 windows with 1-10 tabs each open at all times. Sometimes tabs load indefinitely, sometimes the entire page freezes, as well as puppeteer script execution. Quite often the program will appear to be idle and then open tens of tabs that have apparently been waiting for the newpage command to resolve open at once (usually after clicking into the window, but clicking on the commandline window seems to trigger this as well sometimes). This is incredibly hard to debug, as there seems to be no clear regularity to the issue.

**Comment by gumuruhsspj** [comment]
for everyone who didn't figure it out yet.

Just try to implement these arguments.... and trust me . It works!

```
var ops = {args:[
		'--kiosks',
		'--disable-background-timer-throttling',
		'--disable-backgrounding-occluded-windows',
		'--disable-renderer-backgrounding',
		'--disable-canvas-aa',
        '--disable-2d-canvas-clip-aa',
        '--disable-gl-drawing-for-tests',
        '--disable-dev-shm-usage', 
        '--no-zygote', 
        '--use-gl=desktop', 
        '--enable-webgl',
        '--hide-scrollbars',
        '--mute-audio',
		'--start-maximized',
        '--no-first-run',
        '--disable-infobars',
        '--disable-breakpad', 
        '--user-data-dir='+tempFolder, 
        '--no-sandbox',
        '--disable-setuid-sandbox'
	], headless: false, timeout:0	};
puppeteer = require('puppeteer');
	browser = await puppeteer.launch(ops);
    page = await browser.newPage();
```

**Comment by entrptaher** [comment]
Lets find out exactly which flag made it work, since there are lots of flags that will conflict with many different environment and testing solution.

**Comment by azazrehman** [comment]
> for everyone who didn't figure it out yet.
> 
> Just try to implement these arguments.... and trust me . It works!
> 
> ```
> var ops = {args:[
> 		'--kiosks',
> 		'--disable-background-timer-throttling',
> 		'--disable-backgrounding-occluded-windows',
> 		'--disable-renderer-backgrounding',
> 		'--disable-canvas-aa',
>         '--disable-2d-canvas-clip-aa',
>         '--disable-gl-drawing-for-tests',
>         '--disable-dev-shm-usage', 
>         '--no-zygote', 
>         '--use-gl=desktop', 
>         '--enable-webgl',
>         '--hide-scrollbars',
>         '--mute-audio',
> 		'--start-maximized',
>         '--no-first-run',
>         '--disable-infobars',
>         '--disable-breakpad', 
>         '--user-data-dir='+tempFolder, 
>         '--no-sandbox',
>         '--disable-setuid-sandbox'
> 	], headless: false, timeout:0	};
> puppeteer = require('puppeteer');
> 	browser = await puppeteer.launch(ops);
>     page = await browser.newPage();
> ```

Sorry to say  @fgroupindonesia but still no luck with all of these flags...

**Comment by anubhavmalik** [comment]
If possible please use Headless=true, this would launch the browser without UI and tabs losing focus won't be a problem anymore. 
^ @azazrehman

**Comment by chocofoxy** [comment]
mayeb it's a header issue i fixed this with this code from an article in [medium](https://medium.com/@addnab/puppeteer-quick-fix-for-differences-between-headless-and-headful-versions-of-a-webpage-5b168bd5fe4a)
```
const browser = await puppeteer.launch();
const page = await browser.newPage();

const headlessUserAgent = await page.evaluate(() => navigator.userAgent);
const chromeUserAgent = headlessUserAgent.replace('HeadlessChrome', 'Chrome');
await page.setUserAgent(chromeUserAgent);
await page.setExtraHTTPHeaders({
  'accept-language': 'en-US,en;q=0.8'
});
```

**Comment by x00** [comment]
@sihamza is that not the opposite problem though? This is is not headless, where the there is no window.

**Comment by berteotti** [comment]
+1

**Comment by luucasbraga** [comment]
+1 with the same problem

**Comment by orassayag** [comment]
Have the same issue.. the year is 2021 and 3 years had pasted since the problem declared... any solution to this?

**Comment by liteCarma** [comment]
> 
> 
> Have the same issue.. the year is 2021 and 3 years had pasted since the problem declared... any solution to this?

I also have this problem. If you minimize the browser window, the script pauses

**Comment by ghost** [comment]
🥺👉👈

**Comment by acenturyandabit** [comment]
Multiple browsers, all headless, works for me! I've only got 4 pages though so good luck with >10 haha. Hint: `Promise.all` is your friend :D

**Comment by devinoue** [comment]
+1 with the same problem😭

**Comment by elioth-coder** [comment]
Hi guys I think the solution for this is to just use `page.bringToFront()` to refocus on that page so puppeteer can work correctly.

**Comment by tjaycodes4you** [comment]
+1 same issue.

**Comment by shubham242k** [comment]
Its been almost 3 years and still no luck!!

**Comment by 1nVitr0** [comment]
This is a chromium issue (chromium would probably call it feature). When a tab is not in the foreground, script execution is paused until the tab is foregrounded again. There are no flags that I can confirm working to solve this issue.

The only options you have is either use chromium in headless mode or use `page.bringToFront()` every time you want to do something with the page. This includes waiting for stuff to happen and loading the page. I would definitely recommend using puppeteer in headless mode and if you need visuals for debugging using the chrome://inspect interface

**Comment by jtouris** [comment]
**[Solution]**

Hi everybody, 

I am currently creating an automation that requires multithreading and had this same exact problem. Read through the entire thread and tested several solutions until I was able to fix it. Their was ONE that actually worked for me and was the following:

_Inside of each page that is being opened on a different tab, this code should be invoked:_

```
await page.goto(..............);
await this.wait(4000);
await page.bringToFront()
```
**----  UNFREEZE -------**
`await this.waitForSelector(.......`
- Continue with logic - 

I had several tabs executing at the same time and all of them got unfreeze by the bringToFront event before actually waiting for a specific selector. This has worked for me, hope this helps 

J

**Comment by LukasGirsch** [comment]
@jtourisNS I'am not sure that this is actually solving the underlying issue that puppeteer is just executing the tab in the front. All other tabs are basically paused and waiting their turn.

Just set the browser to headless and the issue is gone. No need to unfreeze and when you are working with multiple pages at the same time there probably is no need to see them anyways.

**Comment by jtouris** [comment]
@LuKaa In my particular scenario, I need the automation **NOT** to be headless due to some logins into specific datacenters with dynamic credentials. I understand that headless mode makes this "easier" but when u do need to see the browser, it is useful to have a workaround that fulfills the needs, right?

My post solution **DID** solved the freezing problem for non-headless automations, once the tab triggers & unfreezes, the rest continues its execution correctly, tabs don't "wait for their turn" anymore and most important, the automation does not crash.

**Comment by fherryfherry** [comment]
I'm facing same issue, I'm using like this solution for that (solved for me) 🤣 like @jtourisNS
```js
await page.goto('https://web.whatsapp.com/',{
  waitUntil: 'networkidle2',
  timeout: 30000
});

// other code

// Forcing page always on front
while(true) {
page.bringToFront();

await delay(1000);
}

```

*you need create `delay()` function by your self .

**Comment by 1nVitr0** [comment]
I have switched to headless mode for a while now and haven't regretted it. If you need visuals just use the same remote debugging protocol that puppeteer uses. `chrome://inspect` is not even strictly needed, it works in all browsers by just visiting your port `http://localhost:9222`. When specifying `0.0.0.0` as your `remoteDebuggingAddress` it even works remotely.

If you still feel you need headful mode, before the switch I used a `requestFocus ` model:

```Javascript
const focusQueue = [];
let queueStarted = false;

function requestFocus(page) {
    return new Promise(resolve => {
        focusQueue.push(() => {
            queueStarted = true;
            page.bringToFront();
            resolve(dequeue);
        });
        // do first dequeue if queue was empty
        if (!queueStarted) dequeue();
    });
}

function dequeue() {
    const nextFocus = focusQueue.shift();
    if (nextFocus) nextFocus();
    else queueStarted = false;
} 
```

The you can use it in your code as:

```javascript
const releaseFocus = await requestFocus(page);

page.type('hello world');
// do stuff with the page

releaseFocus();
```

**Comment by wonghang** [comment]
Actually, I am using [pyppeteer2](https://github.com/dvionst/pyppeteer2)
I have a daemon that uses puppeteer to control a website, and now I have to control 2 websites altogether.

Although I can use `page.bringToFront()` or create two different browser context (https://miyakogi.github.io/pyppeteer/reference.html#browsercontext-class) to work around,
it would be great if I could make everything work using tab, is there any good news on this issue? It has been 3.5 years...

**Comment by rovast** [comment]
Same problem with `"puppeteer-core": "^13.5.2"`, and launch params:

```typescript
const browser: Browser = await puppeteer.launch({
    executablePath: "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
    headless: false,
    userDataDir: config.userDataDir,
    defaultViewport: null
})
```

**Comment by albanx** [comment]
THis issue still happens on Windows, I have facing it right now

**Comment by albanx** [comment]
Ok, just updated to puppeteer 19.x the issue does not happen anymore seems fixed.

**Comment by luiz00martins** [comment]
Still having this issue

**Comment by nguyenvuh2r** [comment]
> ### [Solution]:
> Add the following values to the array you assign to the args property in the options object. Now, you should be able to run Puppeteer while your window is in the foreground, background, or minimized.
> 
> **Update:** I'm running parallel instances of Puppeteer using the same script in multiple windows with one tab per window. Each window completed successfully whether it was in the foreground, background, or minimized. And, it worked in both headless and headful mode.
> 
> ```
> const chromeArgs = [
>   '--disable-background-timer-throttling',
>   '--disable-backgrounding-occluded-windows',
>   '--disable-renderer-backgrounding'
> ];
> ```
> 
> Solution and explanation available in an article written by **Rohit Kumar**, @rkr090 at Big Binary:
> 
> [Debugging failing tests in puppeteer because of background tab](https://blog.bigbinary.com/2018/08/15/debugging-failing-tests-in-background-tab-in-puppeteer.html)

Thanks. It's worked for me!

**Comment by maxktz** [comment]
# SOLUTION

Since chrome has built in feature to emulate a focused page, I am just activating it using CSP session.

Works fine:
```js
const session = await page.createCDPSession();
await session.send(`Emulation.setFocusEmulationEnabled`, { enabled: true });
```

![image](https://github.com/user-attachments/assets/7859b6cd-6510-44b0-a359-02ba73156b52)

**Comment by tsy-ctrl** [comment]
> # SOLUTION
> Since chrome has built in feature to emulate a focused page, I am just activating it using CSP session.
> 
> Works fine:
> 
> ```js
> const session = await page.createCDPSession();
> await session.send(`Emulation.setFocusEmulationEnabled`, { enabled: true });
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/109878406/358569163-7859b6cd-6510-44b0-a359-02ba73156b52.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2NTEzNjUsIm5iZiI6MTcyOTY1MTA2NSwicGF0aCI6Ii8xMDk4Nzg0MDYvMzU4NTY5MTYzLTc4NTliNmNkLTY1MTAtNDRiMC1hMzU5LTAyYmE3MzE1NmI1Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyM1QwMjM3NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYjRhNGNkZjgwNGQxYTEwZDIwOTkzY2YxY2RkN2I5NjVhYTFmN2U4MmNjNTAwMjgzY2QxNDczM2UzYTEzNGNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UMEQa60KuypwZGO6tIiJ80VMYyhb2HnoKMGLqYlwKgw)

ty, works fine for me

**Comment by valkalon** [comment]
Now the question: how to do the same for firefox?
